### PR TITLE
Add intersection analysis

### LINKED ISSUE #7979: Add Intersection analysis
screen shot 2016-06-21 at 09 59 56

screen shot 2016-06-21 at 10 00 09

### PR DESCRIPTION
**Pull request by javierarce**
Closes #7979

### CODE REVIEW

**Comment by javierarce** [comment]
CR: @xavijam 
CC: @javisantana

**Comment by xavijam** [comment]
LGTM 👍 , but I feel @viddo should check changes.

**Comment by javierarce** [comment]
@rochoa: I've discovered a problem when creating an 'aggregate-intersection' analysis. 

I've done an intersection analysis between a polygon and a point layer.

Then I've switcht to aggregation-intersection and selected the 'cartodb_id' column from `b0`, with this result:

```
 Request to Maps API failed: Postgis Plugin: ERROR:  transform: couldn't project point (180 -90 0): tolerance condition error (-20)

in executeQuery Full sql was: 'SELECT ST_AsBinary(ST_Simplify(ST_SnapToGrid("the_geom_webmercator",3.50861e-06), 7.01722e-06)) AS geom FROM (SELECT ST_Transform(the_geom, 3857) the_geom_webmercator, cartodb_id, fid, scalerank, featurecla, labelrank, sovereignt, sov_a3, adm0_dif, level, type, admin, adm0_a3, geou_dif, geounit, gu_a3, su_dif, subunit, su_a3, brk_diff, name, name_long, brk_a3, brk_name, brk_group, abbrev, postal, formal_en, formal_fr, note_adm0, note_brk, name_sort, name_alt, mapcolor7, mapcolor8, mapcolor9, mapcolor13, pop_est, gdp_md_est, pop_year, lastcensus, gdp_year, economy, income_grp, wikipedia, fips_10_, iso_a2, iso_a3, iso_n3, un_a3, wb_a2, wb_a3, woe_id, woe_id_eh, woe_note, adm0_a3_is, adm0_a3_us, adm0_a3_un, adm0_a3_wb, continent, region_un, subregion, region_wb, name_len, long_len, abbrev_len, tiny, homepart, sum_cartodb_id
FROM (WITH
_cdb_analysis_source AS (
 SELECT * FROM librerias_pos_1
),
_cdb_analysis_target AS (
 SELECT * FROM world_borders_hd
)
SELECT
  _cdb_analysis_target.cartodb_id, _cdb_analysis_target.the_geom, _cdb_analysis_target.the_geom_webmercator, _cdb_analysis_target.fid, _cdb_analysis_target.scalerank, _cdb_analysis_target.featurecla, _cdb_analysis_target.labelrank, _cdb_analysis_target.sovereignt, _cdb_analysis_target.sov_a3, _cdb_analysis_target.adm0_dif, _cdb_analysis_target.level, _cdb_analysis_target.type, _cdb_analysis_target.admin, _cdb_analysis_target.adm0_a3, _cdb_analysis_target.geou_dif, _cdb_analysis_target.geounit, _cdb_analysis_target.gu_a3, _cdb_analysis_target.su_dif, _cdb_analysis_target.subunit, _cdb_analysis_target.su_a3, _cdb_analysis_target.brk_diff, _cdb_analysis_target.name, _cdb_analysis_target.name_long, _cdb_analysis_target.brk_a3, _cdb_analysis_target.brk_name, _cdb_analysis_target.brk_group, _cdb_analysis_target.abbrev, _cdb_analysis_target.postal, _cdb_analysis_target.formal_en, _cdb_analysis_target.formal_fr, _cdb_analysis_target.note_adm0, _cdb_analysis_target.note_brk, _cdb_analysis_target.name_sort, _cdb_analysis_target.name_alt, _cdb_analysis_target.mapcolor7, _cdb_analysis_target.mapcolor8, _cdb_analysis_target.mapcolor9, _cdb_analysis_target.mapcolor13, _cdb_analysis_target.pop_est, _cdb_analysis_target.gdp_md_est, _cdb_analysis_target.pop_year, _cdb_analysis_target.lastcensus, _cdb_analysis_target.gdp_year, _cdb_analysis_target.economy, _cdb_analysis_target.income_grp, _cdb_analysis_target.wikipedia, _cdb_analysis_target.fips_10_, _cdb_analysis_target.iso_a2, _cdb_analysis_target.iso_a3, _cdb_analysis_target.iso_n3, _cdb_analysis_target.un_a3, _cdb_analysis_target.wb_a2, _cdb_analysis_target.wb_a3, _cdb_analysis_target.woe_id, _cdb_analysis_target.woe_id_eh, _cdb_analysis_target.woe_note, _cdb_analysis_target.adm0_a3_is, _cdb_analysis_target.adm0_a3_us, _cdb_analysis_target.adm0_a3_un, _cdb_analysis_target.adm0_a3_wb, _cdb_analysis_target.continent, _cdb_analysis_target.region_un, _cdb_analysis_target.subregion, _cdb_analysis_target.region_wb, _cdb_analysis_target.name_len, _cdb_analysis_target.long_len, _cdb_analysis_target.abbrev_len, _cdb_analysis_target.tiny, _cdb_analysis_target.homepart,
  sum(_cdb_analysis_source.cartodb_id)
   as sum_cartodb_id
FROM _cdb_analysis_source JOIN _cdb_analysis_target
  ON ST_Intersects(_cdb_analysis_target.the_geom, _cdb_analysis_source.the_geom)
GROUP BY
  _cdb_analysis_target.cartodb_id, _cdb_analysis_target.the_geom, _cdb_analysis_target.the_geom_webmercator, _cdb_analysis_target.fid, _cdb_analysis_target.scalerank, _cdb_analysis_target.featurecla, _cdb_analysis_target.labelrank, _cdb_analysis_target.sovereignt, _cdb_analysis_target.sov_a3, _cdb_analysis_target.adm0_dif, _cdb_analysis_target.level, _cdb_analysis_target.type, _cdb_analysis_target.admin, _cdb_analysis_target.adm0_a3, _cdb_analysis_target.geou_dif, _cdb_analysis_target.geounit, _cdb_analysis_target.gu_a3, _cdb_analysis_target.su_dif, _cdb_analysis_target.subunit, _cdb_analysis_target.su_a3, _cdb_analysis_target.brk_diff, _cdb_analysis_target.name, _cdb_analysis_target.name_long, _cdb_analysis_target.brk_a3, _cdb_analysis_target.brk_name, _cdb_analysis_target.brk_group, _cdb_analysis_target.abbrev, _cdb_analysis_target.postal, _cdb_analysis_target.formal_en, _cdb_analysis_target.formal_fr, _cdb_analysis_target.note_adm0, _cdb_analysis_target.note_brk, _cdb_analysis_target.name_sort, _cdb_analysis_target.name_alt, _cdb_analysis_target.mapcolor7, _cdb_analysis_target.mapcolor8, _cdb_analysis_target.mapcolor9, _cdb_analysis_target.mapcolor13, _cdb_analysis_target.pop_est, _cdb_analysis_target.gdp_md_est, _cdb_analysis_target.pop_year, _cdb_analysis_target.lastcensus, _cdb_analysis_target.gdp_year, _cdb_analysis_target.economy, _cdb_analysis_target.income_grp, _cdb_analysis_target.wikipedia, _cdb_analysis_target.fips_10_, _cdb_analysis_target.iso_a2, _cdb_analysis_target.iso_a3, _cdb_analysis_target.iso_n3, _cdb_analysis_target.un_a3, _cdb_analysis_target.wb_a2, _cdb_analysis_target.wb_a3, _cdb_analysis_target.woe_id, _cdb_analysis_target.woe_id_eh, _cdb_analysis_target.woe_note, _cdb_analysis_target.adm0_a3_is, _cdb_analysis_target.adm0_a3_us, _cdb_analysis_target.adm0_a3_un, _cdb_analysis_target.adm0_a3_wb, _cdb_analysis_target.continent, _cdb_analysis_target.region_un, _cdb_analysis_target.subregion, _cdb_analysis_target.region_wb, _cdb_analysis_target.name_len, _cdb_analysis_target.long_len, _cdb_analysis_target.abbrev_len, _cdb_analysis_target.tiny, _cdb_analysis_target.homepart) _cdb_analysis_query) as cdbq WHERE "the_geom_webmercator" && ST_MakeEnvelope(-20037508.3,20037508.25881302,-20037508.25881302,20037508.3,3857) LIMIT 65535'
```

**Comment by javierarce** [comment]
@rochoa: an a question (maybe related with my previous comment). Should the aggregation column belong to the source or to the target column?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/intersection-form-model.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    this.listenTo(this._columnOptions, 'columnsFetched', this._setSchema);
+
+    this.on('change:type', this._setSchema, this);
+
+    this._setSchema();
+  },
+
+  getTemplate: function () {
+    if (this.get('type') === 'aggregate-intersection') {
```
Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/intersection-form-model.js]: From what I can tell, the templates are all alike except for the `data-fields` attribute, if that's the case you can have a single template and override `getTemplateData` instead to return the variable data to the template:

```js
// here:
getTemplate: function () {
  return template;
},
getTemplateData: function () {
  return {
    dataFields: this.get('type') === 'aggregate-intersection'
      ? 'type,aggregate_function,aggregate_column'
      : 'type'
  }
}

// template.tpl:
 <div class="Editor-HeaderInfo-inner CDB-Text" data-fields="<%- dataFields %>">
```

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/intersection-form-model.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        type: 'Radio',
+        text: _t('editor.layers.analysis-form.source'),
+        options: [
+          { label: 'Filter', val: 'intersection' },
+          { label: 'Aggregate', val: 'aggregate-intersection' }
+        ]
+      }
+    };
+
+    if (this.get('type') === 'aggregate-intersection') {
```
Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/intersection-form-model.js]: Related to [prev comment](https://github.com/CartoDB/cartodb/pull/8000/files#r67876886), if you're filtering on data-fields you don't need to do it again here

Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/intersection-form-model.js]: Also, if the model represents more than a single type and the parameters aren't the same you might want to override `_formatAttrs` to not pass attrs that won't be used for the analysis node; if they aren't defined in the camshaft-reference they would be persisted as `options` on the model and you probably don't want that.

Comment (javierarce) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-form-models/intersection-form-model.js]: Thanks for the comments, @viddo.

**Comment by dgaubert** [comment]
@javierarce the issue comes even before, it comes from the transformation of geometries that the first analysis returns. Dataset `world_borders_hd` has a geometry of Antarctica that makes the query to fail. This will be fixed when CartoDBfycation of analyses is done. Meanwhile, a workaround would be to avoid that geometry: 

```
select * from world_borders_hd where cartodb_id <> 12;
```

hope this helps!

**Comment by javierarce** [comment]
Thanks, @dgaubert!

**Comment by saleiva** [comment]
So looking fwd to this!

**Pull request merged**